### PR TITLE
identity: abstract identity providers by type

### PR DESCRIPTION
**Pull request by desimone**
## Summary

These changes start to segment the types of authenticating identity providers by type (oidc vs pure oauth2) with an eye on being able to abstract what I'll call the identity augmenting portions of the flow (tacking on group data via a follow on rest call) and for adding support for things like SAML. 

Functionally, this should be identical to the previous version. 

**Checklist**:
- [x] ready for review

### CODE REVIEW

File: internal/identity/oidc/gitlab/gitlab.go [hunk 1]
```diff
@@ -8,12 +9,20 @@ import (
...
 )
 
+var _ identity.Authenticator = &GitLabProvider{}
+
+// Name identifies the GitLab identity provider
+const Name = "gitlab"
+
 const (
 	defaultGitLabProviderURL = "https://gitlab.com"
 	revokeURL                = "https://gitlab.com/oauth/revoke"
```
Comment (Lumexralph) [inline internal/identity/oidc/gitlab/gitlab.go]: adding the `Name` const here

File: internal/identity/oidc/gitlab/gitlab.go [hunk 2]
```diff
@@ -8,12 +9,20 @@ import (
...
+	pom_oidc "github.com/pomerium/pomerium/internal/identity/oidc"
 	"github.com/pomerium/pomerium/internal/sessions"
 	"github.com/pomerium/pomerium/internal/version"
-	"golang.org/x/oauth2"
 )
 
+var _ identity.Authenticator = &GitLabProvider{}
+
+// Name identifies the GitLab identity provider
+const Name = "gitlab"
```
Comment (Lumexralph) [inline internal/identity/oidc/gitlab/gitlab.go]: @desimone Good job! What do you think about adding `Name` into the `consts` declared below [here](https://github.com/pomerium/pomerium/pull/560/files#r398806555), to help improve readability. What do you think?

File: internal/identity/oidc/azure/azure.go [hunk 3]
```diff
@@ -8,15 +10,22 @@ import (
...
+	"github.com/pomerium/pomerium/internal/identity"
+	"github.com/pomerium/pomerium/internal/identity/oidc"
 	"github.com/pomerium/pomerium/internal/log"
 	"github.com/pomerium/pomerium/internal/sessions"
 	"github.com/pomerium/pomerium/internal/version"
 )
 
+var _ identity.Authenticator = &AzureProvider{}
+
+// Name identifies the Azure identity provider
```
Comment (Lumexralph) [inline internal/identity/oidc/azure/azure.go]: @desimone, thinking we can do what I suggested here too to make our code consistent.
https://github.com/pomerium/pomerium/pull/560/files#r398807570

File: internal/identity/oidc/azure/azure.go [hunk 4]
```diff
@@ -8,15 +10,22 @@ import (
...
 
 	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/identity"
+	"github.com/pomerium/pomerium/internal/identity/oidc"
 	"github.com/pomerium/pomerium/internal/log"
 	"github.com/pomerium/pomerium/internal/sessions"
 	"github.com/pomerium/pomerium/internal/version"
 )
 
+var _ identity.Authenticator = &AzureProvider{}
```
Comment (Lumexralph) [inline internal/identity/oidc/azure/azure.go]: @desimone Please be kind to help me understand why we need to do this.

File: internal/identity/oidc/azure/azure.go [hunk 5]
```diff
@@ -8,15 +10,22 @@ import (
...
+
+// Name identifies the Azure identity provider
+const Name = "azure"
```
Comment (Lumexralph) [inline internal/identity/oidc/azure/azure.go]: @desimone, thinking we can do what I suggested here too to make our code consistent.
https://github.com/pomerium/pomerium/pull/560/files#r398807570

File: internal/identity/oidc/azure/azure.go (hunk 4 above)
Comment (desimone) [reply internal/identity/oidc/azure/azure.go]: Ah, for sure. This is a trick to force compile time assertions that an interface is satisfied. In particular, that AzureProvider implements Authenticator. 

https://commaok.xyz/post/compile-time-assertions/

File: internal/identity/oidc/gitlab/gitlab.go (hunk 2 above)
Comment (desimone) [reply internal/identity/oidc/gitlab/gitlab.go]: Good idea.

File: internal/identity/oidc/azure/azure.go (hunk 4 above)
Comment (Lumexralph) [reply internal/identity/oidc/azure/azure.go]: Thanks.

File: internal/identity/oidc/google/google.go [hunk 6]
```diff
@@ -14,58 +15,63 @@ import (
...
 	"github.com/pomerium/pomerium/internal/sessions"
 	"github.com/pomerium/pomerium/internal/version"
 )
 
+var _ identity.Authenticator = &GoogleProvider{}
+
+// Name identifies the Google identity provider
+const Name = "google"
+
 const defaultGoogleProviderURL = "https://accounts.google.com"
```
Comment (Lumexralph) [inline internal/identity/oidc/google/google.go]: based on our discussion above, we can still apply it here https://github.com/pomerium/pomerium/pull/560/files#r398807570

File: internal/identity/oidc/onelogin/onelogin.go [hunk 7]
```diff
@@ -12,62 +13,66 @@ import (
...
 	"github.com/pomerium/pomerium/internal/version"
 )
 
+var _ identity.Authenticator = &OneLoginProvider{}
+
+// Name identifies the OneLogin identity provider
+const Name = "onelogin"
+
 const defaultOneLoginProviderURL = "https://openid-connect.onelogin.com/oidc"
 const defaultOneloginGroupURL = "https://openid-connect.onelogin.com/oidc/me"
```
Comment (Lumexralph) [inline internal/identity/oidc/onelogin/onelogin.go]: @desimone very nice work so far! we can implement this suggestion here too https://github.com/pomerium/pomerium/pull/560/files#r398815591

File: authenticate/authenticate.go [hunk 8]
```diff
@@ -191,3 +199,30 @@ func New(opts config.Options) (*Authenticate, error) {
...
+	case onelogin.Name:
+		a, err = onelogin.New(ctx, &o)
+	default:
+		return nil, fmt.Errorf("authenticate: unknown provider: %s", o.ProviderName)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return a, nil
+}
```
Comment (Lumexralph) [inline authenticate/authenticate.go:228]: Just thoughts: not sure if it'll improve anything, would you be kind to help me know if there was any consideration for putting the `NewIdentityProvider ` function in the `authenticate` or in the `identity` package?

File: internal/identity/oidc/gitlab/gitlab.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	pom_oidc "github.com/pomerium/pomerium/internal/identity/oidc"
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"github.com/pomerium/pomerium/internal/version"
+)
+
+// Name identifies the GitLab identity provider
+const Name = "gitlab"
+
+var defaultScopes = []string{oidc.ScopeOpenID, "api", "read_user", "profile", "email"}
```
Comment (Lumexralph) [inline internal/identity/oidc/gitlab/gitlab.go:25]: update: we can change the `api` scope to `read_api` now based on the new release from gitlab
https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#limiting-scopes-of-a-personal-access-token

Comment (desimone) [reply internal/identity/oidc/gitlab/gitlab.go:25]: 👍 let's fix in a FUP pr so we can reference https://gitlab.com/gitlab-org/gitlab/-/merge_requests/28944

Comment (Lumexralph) [reply internal/identity/oidc/gitlab/gitlab.go:25]: That's cool!

**Pull request merged**